extensions/mssql/src/objectExplorerNodeProvider/hdfsCommands.ts

```diff
@@ -305,11 +305,13 @@ export class PreviewFileCommand extends ProgressCommand {
 				await this.executeWithProgress(
 					async (cancelToken: vscode.CancellationTokenSource) => {
 						let contents = await fileNode.getFileContentsAsString(PreviewFileCommand.DefaultMaxSize);
-						let doc = await this.openTextDocument(fspath.basename(fileNode.hdfsPath));
-						let editor = await this.apiWrapper.showTextDocument(doc, vscode.ViewColumn.Active, false);
-						await editor.edit(edit => {
-							edit.insert(new vscode.Position(0, 0), contents);
-						});
+						let language = this.getLanguage(fileNode.hdfsPath);
+						let doc = await this.apiWrapper.openTextDocument({ language:language, content: contents });
```

Comment:
This loses the filename, and the default of language to plaintext will lose behavior for handlers of .csv, .py or other files. 
- Can you include filename to be used in the new options? This will avoid having it show as untitled
- Can you set language if you've identified it as .sql or .ipynb? This will ensure that these special cases will work in our editors, but default case will also work.

Reply:
It is impossible to use the filename when creating `vscode.TextDocument` with content.
These are the vscode APIs that create `vscode.TextDocument`.
```
export function openTextDocument(uri: Uri): Thenable<TextDocument>;
export function openTextDocument(fileName: string): Thenable<TextDocument>;
export function openTextDocument(options?: { language?: string; content?: string; }): Thenable<TextDocument>;
```

### Why Not Just Use URI or fileName?

If we use `URI` or `filename` for creating `TextDocument`,it will create empty document, and we need to create editor to write the string content into the empty doc. Surprisingly, there is no way to just set the content value of the `TextDocument` without dealing with editor, which turns this issue to headache. 
Writing `.ipynb` file content (string type) into the empty  `TextDocument` by `NotebookEditor` is nightmare. It is not something quick because `NotebookEditor` does not accept straight `string` for editing. Following shows editing APIs in `NotebookEditor`
```
export interface NotebookEditorEdit {
    replace(location: number | CellRange, value: ICellContents): void;
    insertCell(value: ICellContents, index?: number): void;
    deleteCell(index: number): void;
}
```
In order to write the string type content into the `TextDocument`, we need to parse the string content into `ICellContents` object first. Another frustration comes up here: there is no way to parse metadata information in the string content because we do not have metadata object type defined in extension host level, and extension host does not have direct access to metadata object type defined in `vscode` namespace.


### So, no way to set .ipynb string content into named TextDocument directly ?

It is possible. But only in `vscode` level. There is lots of freedom to play with objects and values in `vscode` layer. But problem is that only super limited resource of `vscode` is exposed to extension host level. For this reason, my initial approach was handing over the string content from extension host to `vscode` level since it is impossible to do it in extension host layer.